druid/src/text/editor.rs

```diff
@@ -0,0 +1,311 @@
+    ///
+    /// By default the editor will not wrap lines; this is suitable for
+    /// cases such as a a single-line text field, where the containing
+    /// widget will scroll the editor as required.
+    pub fn set_wrap_width(&mut self, width: f64) {
+        self.layout.set_wrap_width(width);
+    }
+
+    //TODO discussion: should this deref/deref_mut?
+    /// Return the inner [`TextLayout`] objects.
+    ///
+    /// [`TextLayout`]: struct.TextLayout.html
+    pub fn layout(&mut self) -> &mut TextLayout<T> {
+        &mut self.layout
+    }
+}
+
+impl<T: TextStorage + EditableText> Editor<T> {
+    /// Set the text for this editor.
+    ///
+    /// This must be set before the editor is used, such as in [`WidgetAdded`].
+    ///
+    /// [`WidgetAdded`]: ../enum.LifeCycle.html#variant.WidgetAdded
+    pub fn set_text(&mut self, text: T) {
+        self.layout.set_text(text)
+    }
+
+    /// Return the current selection.
+    pub fn selection(&self) -> &Selection {
+        &self.selection
+    }
+
+    /// Returns the `Rect`s representing the  current selection.
+    pub fn selection_rects(&self) -> Vec<Rect> {
+        self.layout.rects_for_range(self.selection.range())
+    }
+
+    /// Returns the `Line` to draw for the current cursor position.
+    pub fn cursor_line(&self) -> Line {
+        self.layout
+            .cursor_line_for_text_position(self.selection.end)
+    }
+
+    /// Handle a mouse click
+    pub fn click(&mut self, mouse: &MouseEvent, data: &mut T) {
+        self.do_edit(EditAction::Click(self.mouse_action_for_event(mouse)), data);
+    }
+
+    /// Handle a mouse drag
+    pub fn drag(&mut self, mouse: &MouseEvent, data: &mut T) {
+        self.do_edit(EditAction::Drag(self.mouse_action_for_event(mouse)), data);
+    }
+
+    /// Handle a copy command
+    pub fn copy(&self, data: &mut T) {
+        if !self.data_is_stale(data) {
+            self.set_clipboard()
+        }
+    }
+
+    /// Handle a cut command
+    pub fn cut(&mut self, data: &mut T) {
+        if !self.data_is_stale(data) {
+            self.set_clipboard();
+            self.delete_backward(data);
+        }
+    }
+
+    /// Handle a paste command
+    pub fn paste(&mut self, t: String, data: &mut T) {
+        self.do_edit(EditAction::Paste(t), data)
+    }
+
+    fn mouse_action_for_event(&self, event: &MouseEvent) -> MouseAction {
+        let pos = self.layout.text_position_for_point(event.pos);
+        MouseAction {
+            row: 0,
+            column: pos,
+            mods: event.mods,
+        }
+    }
+
+    /// Update the editor if the data or env has changed.
+    ///
+    /// The widget owning this `Editor` must call this method during its own
+    /// [`update`] call.
+    ///
+    /// [`update`]: ../trait.Widget.html#tymethod.update
+    pub fn update(&mut self, ctx: &mut UpdateCtx, new_data: &T, env: &Env) {
+        if self.data_is_stale(new_data) {
+            self.layout.set_text(new_data.clone());
```

Comment:
Same here. If the textbox is not multi-lined, we should guard against `new_data` containing `\n` characters!

This issue can be seen when a non-multi lined textbox and another multi-lined textbox lens the same data!

Reply:
this is... a problem.

I'm not sure how best to think about this. Piet has no concept of a single-line layout, and trying to validate external data changes is difficult. It isn't clear what we should do when they happen.

I think my (bad) answer to this, for now, is that it is up to the programmer to ensure that newlines are not inserted externally. We could also definitely debug_panic if a newline exists in the initial text. 😒